test/distributed/fsdp/test_fsdp_hybrid_shard.py

```diff
@@ -171,68 +162,137 @@ def test_fsdp_hybrid_shard_basic_setup(self):
+                self.assertEqual(init_mode, "mixed")
+                continue
+            # process_group should be across the node, which is just the
+            # whole world here.
+            self.assertEqual(
+                dist.get_world_size(fsdp_module.process_group),
+                dist.get_world_size(self.process_group),
+            )
+            intra_node_pgs.add(fsdp_module.process_group)
+            inter_node_pg = fsdp_module._inter_node_pg
+            inter_node_pgs.add(inter_node_pg)
+            self.assertEqual(1, dist.get_world_size(inter_node_pg))
+            self.assertFalse(_rank_not_in_group(inter_node_pg))
+            self.assertEqual(hsdp_sharding_strategy, fsdp_module.sharding_strategy)
+        # All fsdp modules should share the same process groups
+        self.assertEqual(1, len(intra_node_pgs))
+        self.assertEqual(1, len(inter_node_pgs))
+
+        orig_ar = dist.all_reduce
+        orig_rs = dist.reduce_scatter_tensor
+
+        def patched_collective(orig_collective, counter, *args, **kwargs):
+            counter[orig_collective] += 1
+            return orig_collective(*args, **kwargs)
+
+        cntr = Counter()
+        patched_allreduce = partial(patched_collective, orig_ar, cntr)
+        patched_reduce_scatter = partial(patched_collective, orig_rs, cntr)
+        with patch_allreduce(patched_allreduce), patch_reduce_scatter(
+            patched_reduce_scatter
+        ):
+            inp = fsdp_model.get_input(device=torch.cuda.current_device())
+            out = fsdp_model(inp[0], inp[1])
+            loss = fsdp_model.get_loss(inp, out)
+            loss.backward()
+
+        if init_mode == "uniform":
+            num_flat_params = len(list(FSDP._fsdp_handles(fsdp_model)))
+            self.assertEqual(num_flat_params, cntr[orig_ar])
+            self.assertEqual(num_flat_params, cntr[orig_rs])
+        elif init_mode == "mixed":
+            num_hsdp_flat_params = len(list(FSDP._fsdp_handles(fsdp_model.transformer)))
+            num_flat_params = len(list(FSDP._fsdp_handles(fsdp_model)))
+            self.assertEqual(num_hsdp_flat_params, cntr[orig_ar])
+            self.assertEqual(num_flat_params, cntr[orig_rs])
+
+    def _init_hsdp_model(
+        self,
+        hsdp_sharding_strategy: ShardingStrategy,
+        init_mode: str,
+    ):
+        if init_mode == "uniform":
+            auto_wrap_policy = ModuleWrapPolicy(
+                {TransformerEncoderLayer, TransformerDecoderLayer},
+            )
+            fsdp_kwargs = {
+                "auto_wrap_policy": auto_wrap_policy,
+                "device_id": torch.cuda.current_device(),
+                "sharding_strategy": hsdp_sharding_strategy,
+            }
+            fsdp_model = TransformerWithSharedParams.init(
+                self.process_group,
+                FSDPInitMode.RECURSIVE,
+                CUDAInitMode.CUDA_BEFORE,
+                fsdp_kwargs,
+            )
+        elif init_mode == "mixed":
```

Comment:
It looks like the difference between "uniform" and "mixed" is whether we mix and match HSDP sharding strategies. Could we better name these to indicate so?

Reply:
I provided better names: `ALL_HYBRID_SHARD` and `MIXED_HYBRID_FULL_SHARD`.